e2e/testcases/otel_collector_test.go

```diff
@@ -0,0 +1,260 @@
+	cmd := fmt.Sprintf("kubectl %s", strings.Join(args, " "))
+	out, err := nt.Shell.Kubectl(args...)
+	if err != nil {
+		nt.T.Logf("failed to run %q: %v\n%s", cmd, err, out)
+		return err
+	}
+
+	entry := strings.Split(string(out), "\n")
+	for _, m := range entry {
+		if strings.Contains(m, errorString) {
+			return fmt.Errorf("failure found in the log of deployment %s, namespace %s: %s", deployment, namespace, m)
+		}
+	}
+	return nil
+}
+
+// Create a new Monitoring service client using application default credentials
+func createGCMClient(ctx context.Context) (*monitoringv2.MetricClient, error) {
+	client, err := monitoringv2.NewMetricClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
+
+// Make a ListTimeSeries request of a specific metric to GCM with specified
+// metricType.
+// Note: metricType in this context is the metric descriptor name, for example
+// "custom.googleapis.com/opencensus/config_sync/apply_operations_total".
+func listMetricInGCM(ctx context.Context, nt *nomostest.NT, client *monitoringv2.MetricClient, startTime time.Time, metricType string) *monitoringv2.TimeSeriesIterator {
+	endTime := time.Now().UTC()
+	req := &monitoringpb.ListTimeSeriesRequest{
+		Name:   "projects/" + *e2e.GCPProject,
+		Filter: `metric.type="` + metricType + `" AND resource.labels.cluster_name="` + nt.ClusterName + `"`,
+		Interval: &monitoringpb.TimeInterval{
+			StartTime: &timestamp.Timestamp{
+				Seconds: startTime.Unix(),
+			},
+			EndTime: &timestamp.Timestamp{
+				Seconds: endTime.Unix(),
+			},
+		},
+		View: monitoringpb.ListTimeSeriesRequest_HEADERS,
+	}
+	return client.ListTimeSeries(ctx, req)
+}
+
+// Validates a metricType from a specific cluster_name can be found within given
+// TimeSeries
+func validateMetricInGCM(nt *nomostest.NT, it *monitoringv2.TimeSeriesIterator, metricType, clusterName string) error {
+	for {
+		resp, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if resp.GetMetric().GetType() == metricType &&
+			resp.GetResource().GetLabels()["cluster_name"] == clusterName {
+			return nil
+		}
+	}
+	return fmt.Errorf("did not find target time series in GCM: type %s, cluster name %s", metricType, nt.ClusterName)
+}
+
+func setAnnotation(client *testkubeclient.KubeClient, annotationKey, annotationValue string, obj client.Object) (bool, error) {
+	annotations := obj.GetAnnotations()
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+	if annotations[annotationKey] == annotationValue {
+		return false, nil
+	}
+	annotations[annotationKey] = annotationValue
+	obj.SetAnnotations(annotations)
+	if err := client.Update(obj); err != nil {
+		return false, fmt.Errorf("failed to apply updated object to cluster: %v", err)
+	}
+	return true, nil
+}
+
+func unsetAnnotation(client *testkubeclient.KubeClient, annotationKey string, obj client.Object) (bool, error) {
+	annotations := obj.GetAnnotations()
+	if annotations != nil {
```

Comment:
nit: could reduce the layers of nested logic by slightly refactoring this (for readability). e.g.
```golang
func unsetAnnotation(client *testkubeclient.KubeClient, annotationKey string, obj client.Object) (bool, error) {
	annotations := obj.GetAnnotations()
    if annotations == nil {
        return false, nil
    }
    if _, exists := annotations[annotationKey]; !exists {
        return false, nil
    }
    delete(annotations, annotationKey)
	obj.SetAnnotations(annotations)
	if err := client.Update(obj); err != nil {
	    return false, fmt.Errorf("failed to apply updated object to cluster: %v", err)
	}
	return true, nil
}
```

Reply:
Done! Thanks for the suggestion.